dxp/sennajs_integration.md

```diff
@@ -0,0 +1,117 @@
+
+These CSS selectors can also be configured by navigating to **Control Panel** > **System Settings** > **Infrastructure** > **Frontend SPA Infrastructure** and changing the values specificed in the **Navigation Exception Selectors** fields.
+
+## Form submission
+
+The form submission will also be handled by the Senna.js `App` instance, unless the form's `action` attribute is set to `get`
+
+## Block list
+
+It's also possible to tell the Senna.js `App` instance which portlets it needs to ignore, in which case it will not be used for navigation or form submission.
+
+Any portlet that is set to not use "Single Application Mode", or any portlet that is not deployed (and not ready) is automatically added to this list.
+
+## Cache
+
+The Senna.js `App` will use caching, if the `Liferay.SPA.cacheExpirationTime` variable's value is greater than `0`.
+By default DXP sets this value to `-1` which means that the cache is disabled, if you want Senna.js cache to never be cleared you can set this value to `0`
+or specify the value (in minutes) after when Senna.js cache should be cleared.
+
+The cache expiration time can also be configured by navigating to **Control Panel** > **System Settings** > **Infrastructure** > **Frontend SPA Infrastructure** and changing the value specified in the **Cache Expiration Time** field.
+
+With caching enabled, once a resource is loaded, it's content is cached in memory and is retrieved later on without any additional request.
+
+## Excluded paths
+
+The `frontend-js-spa-web` module also introduces a property called [`spa.exluced.paths`](https://github.com/liferay/liferay-portal/blob/454d7f6a439dc668a0ce0529cff2cae87ee9a0d5/modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/spa.properties) which, as its name indicates, is a list of paths that should be ignored by the `Senna.js` App instance.
+
+These paths, are visible in the global `Liferay.SPA.excludedPaths` variable.
+
+## URLs
+
+By default Senna.js in DXP will contain the required parameters needed by portlets, specifically `p_p_id`, `p_p_lifecycle`, `p_p_state` and `p_p_auth`.
+
+## How Senna.js works
+
+![Simplified overview of Senna.js](../images/sennajs-overview.png)
+
+Senna.js, introduces the following concepts:
+
+-   **Routes** Determine which path will be controlled by Senna.js and which Screen will handle the content.
+-   **Surface** A DOM element with an `id` attribute, which is where the content will be rendered.
+-   **Screen** A special type of route handler to handle the navigation request.
+
+To simulate a single page application, Senna.js [tracks all links](https://github.com/liferay/senna.js/blob/059022b737de0f0108b7ef3f06f14d7ade3baa94/src/app/App.js#L104) within the Surface. When a link is triggered, Senna.js controls the event by preventing the default behavior to prevent navigation from being made by the browser and then triggers the navigation process.
```

Comment:
I don't see this explained anywhere... how many surfaces do we have in DXP? which one(s)? This is interesting because people might thing this is working in a way when it's working in a different way.

Reply:
Once you get that, I think it's also interesting to add some diagram or explanation of what type of content are we getting back from DXP backend and what do we do with it.

Spoiler alert, we get full documents as if it were full refreshes (not partials), and then there's some things that get filtered or ignored... I think this would be interesting to write down.

Reply:
Yeah, I think that in the concept of Senna.js there was to be more than one Surface but for DXP only one is used during startup, I don't know exactly why. But yes I can document these details more on the DXP side, in fact, it will happen independently of DXP as well.

Reply:
> Yeah, I think that in the concept of Senna.js there was to be more than one Surface but for DXP only one is used during startup, I don't know exactly why.

I only vaguely remember why now... maybe it can be found in DXP's git history, I'll try to see if I can find the specific commits tomorrow and if not, I'll simply share what I remember.

Reply:
> I only vaguely remember why now... maybe it can be found in DXP's git history, I'll try to see if I can find the specific commits tomorrow and if not, I'll simply share what I remember.

Thanks, I'll try to look for something in the commits too.

> Spoiler alert, we get full documents as if it were full refreshes (not partials), and then there's some things that get filtered or ignored... I think this would be interesting to write down.

Yeah, this is an important thing, I'll make that clearer.

Reply:
The only surface I found was `LiferaySurface` but I'll also search in the git history

Reply:
`LiferaySurface` is a class. When the code run, we instantiate and register a number of surfaces which is what's in the diagram above. Don't go looking in the history for code, figure out which surfaces we create and manage in runtime and then check if we did anything different at any other point in time.

Reply:
I probably should have said that `LiferaySurface` is a class, but the only instance of that class being created is [here](https://github.com/liferay/liferay-portal/blob/3ec7c6e5d08fa594303f140c55407f7158ecccb3/modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/liferay/app/App.es.js#L79). 
At one point we changed [this](https://github.com/liferay/liferay-portal/commit/70b1010a4a521b5a4f955ae281c9e5019faa643a) to prevent overwritting the `document.body.id` if it has one.